tests/vxlan/test_vxlan_ecmp.py

```diff
@@ -630,3 +637,408 @@ def test_vxlan_single_endpoint(self, setUp, encap_type):
+
+        except:
+            self.setup[encap_type]['dest_to_nh_map'] = dict(full_map)
+            # Remove the deleted entry alone.
+            if del_needed:
+                del self.setup[encap_type]['dest_to_nh_map'][vnet][tc10_dest]
+            raise
+
+class Test_VxLAN_ecmp_random_hash(Test_VxLAN):
+    def test_vxlan_random_hash(self, setUp, encap_type):
+        self.setup = setUp
+        Logger.info("tc11: set tunnel route 3 to endpoint group c = {c1, c2, c3}. ensure c1, c2, and c3 matches to underlay default route. send 1000 pkt with random hash to route 3's prefix dst")
+
+        # Chose a vnet for testing.
+        vnet = self.setup[encap_type]['vnet_vni_map'].keys()[0]
+
+        # Create a new destination and 3 nhs for it.
+        tc11_new_dest = get_ip_address(af=get_payload_version(encap_type), netid=Destination_Prefix)
+        tc11_new_nhs = []
+        for i in range(3):
+            tc11_new_nhs.append(get_ip_address(af=get_outer_layer_version(encap_type), netid=NextHop_Prefix))
+
+        # the topology always provides the default routes for any ip address.
+        # so it is already taken care of.
+
+        # Map the new dest and nhs.
+        tc11_config = '[\n' + create_single_route(vnet, tc11_new_dest, HOST_MASK[get_payload_version(encap_type)], tc11_new_nhs, "SET") + '\n]'
+        self.setup[encap_type]['dest_to_nh_map'][vnet][tc11_new_dest] = tc11_new_nhs
+
+        # Apply the config in the DUT and verify traffic. The random hash and ECMP check is already taken care of in the VxLAN PTF script.
+        apply_config_in_swss(self.setup['duthost'], tc11_config, "vnet_route_tc11_"+encap_type)
+        self.dump_self_info_and_run_ptf("tc11", encap_type, True, packet_count=1000)
+
+class Test_VxLAN_underlay_ecmp(Test_VxLAN):
+    @pytest.mark.parametrize("ecmp_path_count", [1, 2])
+    def test_vxlan_modify_underlay_default(self, setUp, minigraph_facts, encap_type, ecmp_path_count):
+        self.setup = setUp
+        Logger.info("tc12: modify the underlay default route nexthop/s. send packets to route 3's prefix dst.")
+        # First step: pick one or two of the interfaces connected to t2, and bring them down.
+        # verify that the encap is still working, and ptf receives the traffic.
+        # Bring them back up.
+        # After that, bring down all the other t2 interfaces, other than the ones used in the first step.
+        # This will force a modification to the underlay default routes nexthops.
+
+        all_intfs = list(get_portchannels_to_neighbors(self.setup['duthost'], "T2", minigraph_facts))
+        if not all_intfs:
+            all_intfs = get_ethernet_to_neighbors("T2", minigraph_facts)
+        Logger.info("Dumping T2 link info: {}".format(all_intfs))
+        if not all_intfs:
+            raise RuntimeError("no interface found connected to t2 neighbors. pls check the testbed, aborting.")
+
+        # Backup
+        all_t2_ports = list(self.setup[encap_type]['t2_ports'])
```

Comment:
I am quite confused here, what is the difference between all_intfs and all_t2_ports, and what does 'backup' mean

Reply:
I understand your confusion. Backup is for the script's internal list of T2 facing-interfaces.
I have modified the code to be more streamlined and also restrictive - only the ports which are up should receive the encap packets, and not the ones that are down.